pygmt/src/grdimage.py

```diff
@@ -179,17 +180,25 @@ def grdimage(self, grid, **kwargs):
     >>> fig.show()
     """
     kwargs = self._preprocess(**kwargs)  # pylint: disable=protected-access
-    with Session() as lib:
-        file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
-        with contextlib.ExitStack() as stack:
-            # shading using an xr.DataArray
-            if kwargs.get("I") is not None and data_kind(kwargs["I"]) == "grid":
-                shading_context = lib.virtualfile_from_data(
-                    check_kind="raster", data=kwargs["I"]
-                )
-                kwargs["I"] = stack.enter_context(shading_context)
 
-            fname = stack.enter_context(file_context)
-            lib.call_module(
-                module="grdimage", args=build_arg_string(kwargs, infile=fname)
-            )
+    with GMTTempFile(suffix=".tif") as tmpfile:
+        if hasattr(grid, "dims") and len(grid.dims) == 3:
+            grid.rio.to_raster(raster_path=tmpfile.name)
+            _grid = tmpfile.name
+        else:
+            _grid = grid
```

Comment:
Should maybe put this logic in `lib.virtualfile_from_data`, since we'll need to reuse it for `grdview`, `grdcut`, and possibly other GMT modules that work with GMT_IMAGE.

Reply:
Ok, created a new `tempfile_from_image` function at 89da91603705b817d70bf60f08464ba817f40314 (similar to `tempfile_from_geojson`). This can be replaced with `lib.virtualfile_from_image` once that is implemented.